### PR TITLE
Extract kprobe-based tracer to adhere to common interface

### PR DESCRIPTION
**Pull request by brycekahle**
### What does this PR do?

Extracts the eBPF network tracer to follow an interface that alternative implementations can follow.

### Motivation

This will allow us to easily evaluate other tracer implementations without churning a lot of code.

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

### CODE REVIEW

File: pkg/network/ebpf/kprobe_types.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+//+build ignore
```
Comment (p-lambert) [inline pkg/network/ebpf/kprobe_types.go:1]: 👍

File: pkg/network/ebpf/kprobe_types_linux.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
+// Code generated by cmd/cgo -godefs; DO NOT EDIT.
+// cgo -godefs -- kprobe_types.go
```
Comment (p-lambert) [inline pkg/network/ebpf/kprobe_types_linux.go:2]: awesome!

File: pkg/network/ephemeral_linux.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	ephemeralLow  = uint16(0)
+	ephemeralHigh = uint16(0)
+
+	initEphemeralIntPair sync.Once
+	ephemeralIntPair     *sysctl.IntPair
+)
+
+// IsPortInEphemeralRange returns whether the port is ephemeral based on the OS-specific configuration.
+func IsPortInEphemeralRange(p uint16) EphemeralPortType {
+	initEphemeralIntPair.Do(func() {
```
Comment (p-lambert) [inline pkg/network/ephemeral_linux.go:21]: 👍

File: pkg/network/event_common.go [hunk 4]
```diff
@@ -194,6 +194,8 @@ type ConnectionStats struct {
 	DNSStatsByDomainByQueryType map[*intern.Value]map[dns.QueryType]dns.Stats
 
 	Via *Via
+
+	IsAssured bool
```
Comment (p-lambert) [inline pkg/network/event_common.go:198]: Why was this added?

File: pkg/network/tracer/kprobe/tracer.go [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+
+type ConnectionTracer interface {
+	Start() (<-chan network.ConnectionStats, error)
+	Stop()
+	GetConnections(buffer []network.ConnectionStats, filter func(*network.ConnectionStats) bool) ([]network.ConnectionStats, error)
+	FlushPending() []network.ConnectionStats
+	Remove(conn *network.ConnectionStats) error
+	GetTelemetry() map[string]int64
+	GetMap(string) *ebpf.Map
+}
```
Comment (p-lambert) [inline pkg/network/tracer/kprobe/tracer.go:39]: A couple of thoughts/questions here:
* Shouldn't this `interface` be defined somewhere else (like `pkg/network/tracer`)?
* Given that the interface itself is quite complex, I'd document the interface methods as they serve as a guideline for alternative implementations.
* I worry that some of the implementations details are leaking into the interface. For example, is `FlushPending` really necessary or can we call it from `GetConnections` or something like that?

**Review by p-lambert — COMMENTED** [review]
Left a few comments about the `ConnectionTracer` interface. Other than that I think this is definitely a huge improvement over what we have now. 🙇‍♂️

File: pkg/network/tracer/kprobe/tracer.go (hunk 5 above)
Comment (brycekahle) [reply pkg/network/tracer/kprobe/tracer.go:39]: `GetConnections` is for active connections, whereas `FlushPending` is for already closed connections that haven't been reported to userspace yet.

The problem is that the closed connections normally flow via the channel returned from `Start`. There is no way using that channel to ensure all the pending connections are processed before doing the delta state work.

Comment (brycekahle) [reply pkg/network/tracer/kprobe/tracer.go:39]: > Shouldn't this `interface` be defined somewhere else (like `pkg/network/tracer`)?

Moving to `pkg/network/tracer` would cause an import loop, since the `kprobe.NewTracer` function returns the interface. I could move the interface to its own package, and have the implementations one level deeper, similar to how `hash` is laid out in the stdlib.

File: pkg/network/event_common.go (hunk 4 above)
Comment (brycekahle) [reply pkg/network/event_common.go:198]: This flag comes from the tcp stats map, but is used in determining the timeout for a flow. In order to maintain an abstraction, it needed to be added to `ConnectionStats`.

File: pkg/network/tracer/kprobe/tracer.go (hunk 5 above)
Comment (brycekahle) [reply pkg/network/tracer/kprobe/tracer.go]: > I'd document the methods

Documentation added.

Comment (p-lambert) [reply pkg/network/tracer/kprobe/tracer.go]: > I could move the interface to its own package, and have the implementations one level deeper, similar to how hash is laid out in the stdlib.
I think that would probably make more sense since now the interface resides inside the kprobe package which just happens to be a possible implementation of it.

**Review by p-lambert — APPROVED** [review]
Looks great 🎉
